pkg/services/ngalert/api/api_convert_prometheus.go

```diff
@@ -377,6 +391,116 @@ func (srv *ConvertPrometheusSrv) RouteConvertPrometheusPostRuleGroup(c *contextm
+	logger := srv.logger.FromContext(c.Req.Context())
+
+	pauseRecordingRules := c.QueryBoolWithDefault("pauseRecordingRules", true)
+	pauseAlerts := c.QueryBoolWithDefault("pauseAlerts", true)
+	targetFolderUID := c.Query("targetFolderUID")
+	namespaceFilter := c.Query("namespace")
+	groupFilter := c.Query("group")
+
+	// 1. Fetch rules from datasource
+	promGroups := map[string][]prom.PrometheusRuleGroup{}
+	resp := srv.proxySvc.requester.withReq(c, http.MethodGet, withPath(*c.Req.URL, MimirPrefix), nil, yamlExtractor(&promGroups), nil)
+
+	// 2. Convert Prometheus Rules to GMA
+	if err := json.Unmarshal(resp.Body(), &promGroups); err != nil {
+		return errorToResponse(err)
+	}
+
+	grafanaGroups := make([]*ngmodels.AlertRuleGroup, 0, len(promGroups))
+	for ns, rgs := range promGroups {
+		// Check Namespace filter
+		if namespaceFilter != "" && ns != namespaceFilter {
+			continue
+		}
+
+		// Need to filter groups here so we don't end up unnecessarily creating empty namespaces
+		if groupFilter != "" {
+			filteredRgs := make([]prom.PrometheusRuleGroup, 0, len(rgs))
+			for _, rg := range rgs {
+				if rg.Name != groupFilter {
+					continue
+				}
+				filteredRgs = append(filteredRgs, rg)
+			}
+			rgs = filteredRgs
+		}
+
+		logger.Debug("Creating a new namespace", "title", ns)
+		namespace, errResp := srv.getOrCreateNamespace(c, ns, logger, targetFolderUID)
+		if errResp != nil {
+			logger.Error("Failed to create a new namespace", "folder_uid", targetFolderUID)
+			return errResp
+		}
+
+		ds, err := srv.datasourceCache.GetDatasourceByUID(c.Req.Context(), datasourceUID, c.SignedInUser, c.SkipDSCache)
+		if err != nil {
+			return ErrResp(http.StatusInternalServerError, err, "failed to get datasource by UID")
+		}
+
+		// We don't want to keep the original definition for rules converted in the UI
+		// as they are editable and the stored original rule definition can become stale
+		promConverter, err := prom.NewConverter(
+			prom.Config{
+				DatasourceUID:   ds.UID,
+				DatasourceType:  ds.Type,
+				DefaultInterval: srv.cfg.DefaultRuleEvaluationInterval,
+				RecordingRules: prom.RulesConfig{
+					IsPaused: pauseRecordingRules,
+				},
+				AlertRules: prom.RulesConfig{
+					IsPaused: pauseAlerts,
+				},
+				KeepOriginalRuleDefinition: util.Pointer(false),
+			},
+		)
+		if err != nil {
+			logger.Error("Failed to create Prometheus converter", "datasource_uid", ds.UID, "datasource_type", ds.Type, "error", err)
+			return response.Err(err)
+		}
+
+		for _, rg := range rgs {
+			grafanaGroup, err := promConverter.PrometheusRulesToGrafana(c.SignedInUser.GetOrgID(), namespace.UID, rg)
+			if err != nil {
+				logger.Error("Failed to convert Prometheus rules to Grafana rules", "error", err)
+				return response.Err(err)
+			}
+			grafanaGroups = append(grafanaGroups, grafanaGroup)
+		}
+	}
+
+	// 3. Update the GMA Rules in the DB
+	err := srv.xactManager.InTransaction(c.Req.Context(), func(ctx context.Context) error {
```

Comment:
I do not think we need it here. The first thing `updateAlertRulesInGroup` does is to initiate the transaction

Reply:
The inner transaction works at the group rules level, whereas the outer level works per group. If a group fails to write then I want the entire thing to be rolled back, so I believe we need an outer and inner transaction scope.

Reply:
makes sense